File: leetcode/surrounded-regions.java
// https://leetcode.com/problems/surrounded-regions/

class Solution {
    public void solve(char[][] board) {
        // First and Last columns
        for(int i = 0; i<board.length; i++){
            if(board[i][0] == 'O') markFreeRegions(board, i, 0);
            if(board[i][board[0].length-1] == 'O') markFreeRegions(board, i, board[0].length-1);
        }
        
        // Top and Bottom Rows
        for(int i = 1; i<board[0].length-1; i++){
            if(board[0][i] == 'O') markFreeRegions(board, 0, i);
            if(board[board.length-1][i] == 'O') markFreeRegions(board, board.length-1, i);
        }
        
        // Capture 0s, replace Free places with O
        for(int i = 0; i<board.length; i++){
            for(int j = 0; j<board[0].length; j++){
                if(board[i][j] == 'O') board[i][j] = 'X';
                if(board[i][j] == 'F') board[i][j] = 'O';
            }
        }
    }
    
    public void markFreeRegions(char[][] board, int i, int j){
        if(i >= 0 && j >= 0 && i < board.length && j < board[0].length && board[i][j] == 'O'){
            board[i][j] = 'F';
            markFreeRegions(board, i+1, j);
            markFreeRegions(board, i-1, j);
            markFreeRegions(board, i, j+1);
            markFreeRegions(board, i, j-1);
        }
    }
}

File: leetcode/surrounded_regions.py
# https://leetcode.com/problems/surrounded-regions/description/


class Solution:
    def solve(self, board: list[list[str]]) -> None:
        # Do not return anything, modify board in-place instead.

        # Mark edge-adjacent 0s with E
        # check top and bottom rows for Os
        for c in range(len(board[0])):
            if board[0][c] == "O":
                self.can_escape(board, 0, c)
            if board[len(board) - 1][c] == "O":
                self.can_escape(board, len(board) - 1, c)

        # check left and right columns for Os
        for r in range(1, len(board)):
            if board[r][0] == "O":
                self.can_escape(board, r, 0)
            if board[r][len(board[0]) - 1] == "O":
                self.can_escape(board, r, len(board[0]) - 1)

        # replace all remaining Os with X
        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == "O":
                    board[r][c] = "X"

        # replace E with O
        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == "E":
                    board[r][c] = "O"

    def can_escape(self, board: list[list[str]], row: int, col: int) -> None:
        if (
            row >= 0
            and col >= 0
            and row < len(board)
            and col < len(board[0])
            and board[row][col] == "O"
        ):
            board[row][col] = "E"
            self.can_escape(board, row + 1, col)
            self.can_escape(board, row - 1, col)
            self.can_escape(board, row, col + 1)
            self.can_escape(board, row, col - 1)
